src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala

```diff
@@ -1,23 +1,55 @@
 
 import com.typesafe.config.Config
 import com.typesafe.scalalogging.LazyLogging
-
 import akka.http.scaladsl.model.StatusCodes
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import net.ceedubs.ficus.Ficus._
+import org.broadinstitute.dsde.workbench.leonardo.auth.sam.SamAuthProvider.{CanSeeAllClustersInProjectCacheKey, NotebookAuthCacheKey, SamCacheKey}
 import org.broadinstitute.dsde.workbench.leonardo.model._
 import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterName
 import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions._
 import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions.CreateClusters
 import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
 import org.broadinstitute.dsde.workbench.model.google.GoogleProject
 
+import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
 
 case class UnknownLeoAuthAction(action: LeoAuthAction)
   extends LeoException(s"SamAuthProvider has no mapping for authorization action ${action.toString}, and is therefore probably out of date.", StatusCodes.InternalServerError)
 
+object SamAuthProvider {
+  private[sam] sealed trait SamCacheKey
+  private[sam] case class NotebookAuthCacheKey(userEmail: WorkbenchEmail, action: NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName, executionContext: ExecutionContext) extends SamCacheKey
+  private[sam] case class CanSeeAllClustersInProjectCacheKey(userEmail: WorkbenchEmail, googleProject: GoogleProject, executionContext: ExecutionContext) extends SamCacheKey
+}
+
 class SamAuthProvider(val config: Config, serviceAccountProvider: ServiceAccountProvider) extends LeoAuthProvider(config, serviceAccountProvider) with SamProvider with LazyLogging {
 
+  private lazy val notebookAuthCacheMaxSize = config.getAs[Int]("notebookAuthCacheMaxSize").getOrElse(1000)
+  private lazy val notebookAuthCacheExpiryTime = config.getAs[FiniteDuration]("notebookAuthCacheExpiryTime").getOrElse(15 minutes)
+
+  // Cache notebook auth results from Sam as this is called very often by the proxy and the "list clusters" endpoint.
+  // Project-level auth is not cached because it's just called for "create cluster" which is not as frequent.
+  private[sam] val notebookAuthCache = CacheBuilder.newBuilder()
+    .expireAfterWrite(notebookAuthCacheExpiryTime.toSeconds, TimeUnit.SECONDS)
+    .maximumSize(notebookAuthCacheMaxSize)
+    .build(
+      new CacheLoader[SamCacheKey, Future[Boolean]] {
+        def load(key: SamCacheKey) = {
+          key match {
+            case NotebookAuthCacheKey(userEmail, action, googleProject, clusterName, executionContext) =>
+              hasNotebookClusterPermissionInternal(userEmail, action, googleProject, clusterName)(executionContext)
+            case CanSeeAllClustersInProjectCacheKey(userEmail, googleProject, executionContext) =>
+              canSeeAllClustersInProjectInternal(userEmail, googleProject)(executionContext)
```

Comment:
Cool - I like this way of separating out between the caches - I was imagining 2 separate caches, but this is better i think

Reply:
Yeah it seemed to make sense because we're caching similar things (sam auth results) which have the same type (Boolean).